cmd/oras/discover.go

```diff
@@ -136,19 +139,59 @@ func runDiscover(opts discoverOptions) error {
 	return nil
 }
 
-func fetchReferrers(ctx context.Context, repo *remote.Repository, desc ocispec.Descriptor, artifactType string) ([]ocispec.Descriptor, error) {
-	results := []ocispec.Descriptor{}
-	err := repo.Referrers(ctx, desc, artifactType, func(referrers []ocispec.Descriptor) error {
-		results = append(results, referrers...)
-		return nil
-	})
-	if err != nil {
-		return nil, err
+func fetchReferrers(ctx context.Context, target oras.ReadOnlyGraphTarget, desc ocispec.Descriptor, artifactType string) ([]ocispec.Descriptor, error) {
+	var results []ocispec.Descriptor
+	if repo, ok := target.(*remote.Repository); ok {
+		// get referrers directly
+		err := repo.Referrers(ctx, desc, artifactType, func(referrers []ocispec.Descriptor) error {
+			results = append(results, referrers...)
+			return nil
+		})
+		if err != nil {
+			return nil, err
+		}
+	} else {
+		// find matched referrers in all predecessors
+		predecessors, err := target.Predecessors(ctx, desc)
+		if err != nil {
+			return nil, err
+		}
+		for _, node := range predecessors {
+			var fetched []byte
+			if rc, err := target.Fetch(ctx, node); err != nil {
+				return nil, err
+			} else {
+				fetched, err = content.ReadAll(rc, node)
+				if err != nil {
+					return nil, err
+				}
+			}
+			var isMatch = func(got *ocispec.Descriptor, want ocispec.Descriptor) bool {
+				return got != nil && content.Equal(*got, want)
+			}
+			switch node.MediaType {
+			case ocispec.MediaTypeArtifactManifest:
+				var artifact ocispec.Artifact
+				json.Unmarshal(fetched, &artifact)
+				if isMatch(artifact.Subject, desc) {
+					node.ArtifactType = artifact.ArtifactType
+				}
+			case ocispec.MediaTypeImageManifest:
+				var image ocispec.Manifest
+				json.Unmarshal(fetched, &image)
+				if isMatch(image.Subject, desc) {
+					node.ArtifactType = image.Config.MediaType
+				}
+			}
+			if node.ArtifactType != "" && (artifactType == "" || artifactType == node.ArtifactType) {
+				results = append(results, node)
+			}
+		}
 	}
 	return results, nil
```

Comment:
nit: code style: it would be more readable and less indents to have
```go
var result T
if ... {
    ...
    return result
}

...
return result
```
instead of
```go
if ... {
    ...
} else {
    ...
}

return result
```